Option: Use fused_batch_norm instead of batch_norm for layer normalization

thorjohnsen opened the pull request:
tf.nn.fused_batch_norm uses cudnn for GPU acceleration, tf.nn.batch_normalization does not. This PR adds the option of using fused_batch_norm instead of batch_normalization. It is an opt-in option because fused_batch_norm only improves runtimes on GPUs and does not support double precision tensors.

thorjohnsen committed:
Optionally use fused_batch_norm to perform layer_norm, speeds up GPU runs

AzizCode92 commented:
Thank you for approaching this issue. 
Have you tried this change on a specific task? Did it make the layer normalization faster?

thorjohnsen commented:
@AzizCode92 I used it in the attention layer of the MLPerf Transformer benchmark. It made the benchmark run about 25% faster. I don't know how familiar you are with MLPerf, it is a google initiative, the public repository is in https://github.com/mlperf/training/. The Transformer reference code has its own implementation of layer normalization (lines 247-265 in file translation/tensorflow/transformer/model/transformer.py), I changed that to use tf.contrib.layers.layer_norm, which simplified the code, but did not make it any faster. I added the option to use tf.nn.fused_batch_norm and got a 25% jump in throughput on GPUs when running our mixed precision version of Transformer. I have not tested what kind of impact this has on the reference code, which uses purely fp32 arithmetic.

AzizCode92 commented:
Interesting!!! I'm sorry I have no idea about the task you're working on.
I'm applying your patch to the  tf.contrib.rnn.LayerNormBasicLSTMCell where they're applying the layer_norm function. Before I noticed that applying the current code, the training time was too slow. 
Have a look at [this issue](https://github.com/tensorflow/tensorflow/issues/21929)
I have also to notice that I combine your modification to this interesting [pull request ](https://github.com/tensorflow/tensorflow/pull/15861).
I'm still training my model and I hope with these two patches I can achieve layerNorm faster and more accurate. 
thank you

AzizCode92 commented:
#Update : it is definitely faster

AzizCode92 commented:
I also recommend you to see this [pull request](https://github.com/MycChiu/fast-LayerNorm-TF)

thorjohnsen commented:
@AzizCode92 Thanks for telling me about the custom layer norm PR, I'll give that a try.

AzizCode92 commented:
Good, I have worked on it. Unfortunately all the test cases failed!!!
It is an old repo ( 2 years ago ) but worth spending time to fix it

AzizCode92 commented:
#Update : the latest PR I mentioned is working fine now.
Current tensorflow version : 1.10
Cuda v.9
and I can confirm it is faster than the normal layer_norm operation we have now

fchollet commented on tensorflow/contrib/layers/python/layers/layers.py:
```diff
@@ -2205,7 +2206,8 @@ def layer_norm(inputs,
                begin_params_axis=-1,
-               scope=None):
+               scope=None,
+               use_fused_batch_norm=False):
```
You can just call this argument "fused".

fchollet commented on tensorflow/contrib/layers/python/layers/layers.py:
```diff
@@ -2307,19 +2311,57 @@ def layer_norm(inputs,
+        outputs_shape = tf.shape(inputs)
+      inputs = array_ops.reshape(inputs, bn_shape)
+      if inputs.get_shape().is_fully_defined():
+        # static inputs TensorShape fully defined after reshape.
```
?

fchollet reviewed (commented):
Thanks for the PR.

Please add appropriate unit tests.

Also, note that `contrib.layers` is deprecated and will be removed entirely in TensorFlow 2.0, which is coming soon.

tensorflowbutler commented:
Nagging Assignee @yifeif: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

yifeif commented:
@thorjohnsen could you address @fchollet 's comments? Thanks!

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.
